### PR TITLE
Updates to STX control PR

### PR DESCRIPTION
**Pull request by jdinan**
Updates to the PR currently posted for SOS.

@davidozog Please review and merge changes as appropriate.

### CODE REVIEW

**Comment by jdinan** [comment]
Note: only the last 4 commits are changes to the PR, the rest came from merging master with the branch.

**Review by davidozog — CHANGES_REQUESTED** [review]
This looks great on many levels!

What do you think of doing `want_used` instead of `want_unused`?  It's the weirder case, but I have a hunch it would be ever so slightly easier to pickup for someone who hasn't seen the code before.

**Comment by jdinan** [comment]
Yeah, I've been ruminating on that bit, too.  Here's a rework of that; let me know what you think.

**Comment by jdinan** [comment]
af48954 ❤️ 
```
$ SHMEM_DEBUG=1 SHMEM_OFI_STX_MAX=4 SHMEM_OFI_STX_ALLOCATOR=random mpiexec -np 1 test/unit/shmem_ctx
Sandia OpenSHMEM 1.4.0rc1
Build information:
  Git Version           v1.4.0rc1-41-gad3a93b2-dirty (wip/stx_control)
  Configure Args        '--prefix=/Users/dinanjam/opt/sos-dev' '--enable-pmi-simple'
                        '--enable-thread-completion'
                        '--with-ofi=/Users/dinanjam/opt/libfabric-1.5.1'
                        '--enable-debug' '--enable-error-checking' 'CC=gcc-7'
                        'CXX=g++-7'
  Build Date            Thu Jan 11 21:18:18 EST 2018
  Build CC              gcc-7
  Build CFLAGS          -std=gnu11 -g -O2 -Wall -g -fvisibility=hidden

[0000] DEBUG: ../../src/init.c:247: shmem_internal_init
[0000]        Sym. heap=0x180000000 len=537919488 -- data=0x100000c87 len=9081
[0000] DEBUG: ../../src/transport_ofi.c:1235: query_for_fabric
[0000]        OFI provider: sockets, fabric: 127.0.0.0/8, domain: lo0
[0000] DEBUG: ../../src/transport_ofi.c:1089: allocate_fabric_resources
[0000]        OFI version: built 1.5, cur. 1.5; provider version: 2.0
[0000] DEBUG: ../../src/transport_ofi.c:530: shmem_transport_ofi_stx_allocate
[0000]        STX[4] = [ 1S 0S 0S 0S ]
[0000] DEBUG: ../../src/transport_ofi.c:530: shmem_transport_ofi_stx_allocate
[0000]        STX[4] = [ 1S 1P 0S 0S ]
[0000] DEBUG: ../../src/transport_ofi.c:530: shmem_transport_ofi_stx_allocate
[0000]        STX[4] = [ 1S 1P 1P 0S ]
[0000] DEBUG: ../../src/transport_ofi.c:530: shmem_transport_ofi_stx_allocate
[0000]        STX[4] = [ 1S 1P 1P 1P ]
[0000] DEBUG: ../../src/transport_ofi.c:491: shmem_transport_ofi_stx_allocate
[0000]        private STX unavailable, falling back to STX sharing
[0000] DEBUG: ../../src/transport_ofi.c:530: shmem_transport_ofi_stx_allocate
[0000]        STX[4] = [ 2S 1P 1P 1P ]
[0000] DEBUG: ../../src/transport_ofi.c:491: shmem_transport_ofi_stx_allocate
[0000]        private STX unavailable, falling back to STX sharing
[0000] DEBUG: ../../src/transport_ofi.c:530: shmem_transport_ofi_stx_allocate
[0000]        STX[4] = [ 3S 1P 1P 1P ]
[0000] DEBUG: ../../src/transport_ofi.c:491: shmem_transport_ofi_stx_allocate
[0000]        private STX unavailable, falling back to STX sharing
[0000] DEBUG: ../../src/transport_ofi.c:530: shmem_transport_ofi_stx_allocate
[0000]        STX[4] = [ 4S 1P 1P 1P ]
[0000] DEBUG: ../../src/transport_ofi.c:491: shmem_transport_ofi_stx_allocate
[0000]        private STX unavailable, falling back to STX sharing
[0000] DEBUG: ../../src/transport_ofi.c:530: shmem_transport_ofi_stx_allocate
[0000]        STX[4] = [ 5S 1P 1P 1P ]
[0000] DEBUG: ../../src/transport_ofi.c:491: shmem_transport_ofi_stx_allocate
[0000]        private STX unavailable, falling back to STX sharing
[0000] DEBUG: ../../src/transport_ofi.c:530: shmem_transport_ofi_stx_allocate
[0000]        STX[4] = [ 6S 1P 1P 1P ]
```

File: src/transport_ofi.c [hunk 1]
```diff
@@ -312,179 +313,224 @@ typedef enum{
...
             }
-            e->tid = ctx->tid;
+            e->tid     = ctx->tid;
             e->stx_idx = ctx->stx_idx;
-            e->stx = &shmem_transport_ofi_stx_pool[ctx->stx_idx];
-            e->stx->ref_cnt++;
-            e->stx->is_private = true;
 
-            HASH_ADD_INT(shmem_transport_ofi_stx_kvs, tid, e);
+            if (is_unused) {
```
Comment (davidozog) [inline src/transport_ofi.c:508]: Does the `e` buffer leak if `is_unused == 0`?

File: src/transport_ofi.c [hunk 2]
```diff
@@ -312,179 +313,224 @@ typedef enum{
...
+                stx->is_private = 1;
+                HASH_ADD_INT(shmem_transport_ofi_stx_kvs, tid, e);
+            } else {
+                ctx->options &= ~SHMEM_CTX_PRIVATE;
+            }
         }
+    /* TODO: Optimize this case? else if (ctx->options & SHMEM_CTX_PRIVATE) */
```
Comment (davidozog) [inline src/transport_ofi.c:515]: Not sure I understand this comment... Do you mean `else if (!(ctx->options & SHMEM_CTX_PRIVATE))` with an `else {RAISE_ERROR...}`, or do you mean something else?

File: src/transport_ofi.c [hunk 3]
```diff
@@ -312,179 +313,224 @@ typedef enum{
...
+    } else {
+        int stx_idx = shmem_transport_ofi_stx_search_unused();
 
+        /* TODO: Currently only support greedy assignment of STX to non-private
+         * contexts.  In the future, we may want to set a threshold for refs on
+         * a shared STX before allocating the next one. */
```
Comment (davidozog) [inline src/transport_ofi.c:521]: 👍

File: src/transport_ofi.c [hunk 4]
```diff
@@ -312,179 +313,224 @@ typedef enum{
...
+typedef enum stx_allocator_t stx_allocator_t;
+static stx_allocator_t shmem_transport_ofi_stx_allocator;
 
 static long shmem_transport_ofi_stx_max;
 
 struct shmem_transport_ofi_stx_t {
     struct fid_stx*   stx;
     long              ref_cnt;
-    bool              is_private;
+    int               is_private;
```
Comment (davidozog) [inline src/transport_ofi.c:328]: Any particular reason not to use `<stdbool.h>`?

**Review by davidozog — APPROVED** [review]

**Pull request merged**